A36-xds-for-servers.md

```diff
@@ -0,0 +1,401 @@
+XdsServer should continue using the most recent configuration until connectivity
+is restored. However, XdsServer must accept configuration changes provided by
+the xDS server, including resource deletions. If that causes the XdsServer to
+lack essential configuration (e.g., the Listener was deleted) the server would
+need to enter "not serving" mode. "Not serving" requires existing connections to
+be closed, but already-started RPCs should not fail. The XdsServer is permitted
+to use the previously-existing configuration to service RPCs during a two-phase
+GOAWAY to avoid any RPC failures, or it may use a one-phase GOAWAY which will
+fail racing RPCs but in a way that the client may transparently retry. The
+XdsServer is free to use a different "not serving" strategy post-startup than
+for the initial startup.
+
+The XdsServer does not have to wait until server credentials (e.g., TLS certs)
+are available before accepting connections; since XdsServerCredentials might not
+be used, the server is free to lazily load credentials. However, the XdsServer
+should keep the credentials cache fresh and up-to-date after that initial
+lazy-loading, as it is clear at that point that XdsServerCredentials are being
+used.
+
+The XdsServer API will allow applications to register a "serving state" callback
+to be invoked when the server begins serving and when the server encounters
+errors that force it to be "not serving". If "not serving", the callback must be
+provided error information, for debugging use by developers. The error
+information should generally be language-idiomatic, but determining the cause of
+the error does not need to be machine-friendly. If the application does not
+register the callback, XdsServer should log any errors and each serving
+resumption after an error, all at a default-visible log level.
+
+XdsServer's start must not fail due to transient xDS issues, like missing xDS
+configuration from the xDS server. If XdsServer's start blocks waiting for xDS
+configuration an application can use the serving state callback to be notified
+of issues preventing startup progress.
+
+### xDS Protocol
+
+The `GRPC_XDS_BOOTSTRAP` file will be enhanced to have a new field:
+```
+{
+  // A template for the name of the Listener resource to subscribe to for a gRPC
+  // server. If the token `%s` is present in the string, it will be replaced
+  // with the server's listening "IP:port" (e.g., "0.0.0.0:8080", "[::]:8080").
+  "server_listener_resource_name_template": "example/resource/%s",
+  // ...
+}
+```
+
+XdsServer will use the normal `XdsClient` to communicate with the xDS server.
+There is no default value for `server_listener_resource_name_template` so if it
+is not present in the bootstrap then server creation or start will fail or the
+XdsServer will become "not serving". XdsServer will perform the `%s` replacement
+in the template (if the token is present) to produce a listener name. The
```

Comment:
When we say `if the token is present` (which basically means this template is free-form), how are we going to ensure that gRPC can extract the actual listening_address out of it when we have to match it against the one stored in the `address` field of the received `Listener` resource?
@sanjaypujare @yashykt

Reply:
We don't extract the listening_address out of it. The matching of the address field is handled by the high-level XdsServer, not low-level XdsClient. XdsServer knows which address it is listening on.

Reply:
Thanks. So, the XdsServer replaces the token to generate the resource_name when requesting the Listener resource. But when matching the address field on the received Listener, it uses the address that it knows for sure that it is listening on. That makes sense.